### PR TITLE
Add more utility functions for job and status

### PR DESCRIPTION
**Pull request by jian-he**
Add more utility functions for job and status. 
This echos what's already available in TF operator.

@terrytangyuan 
@richardsliu 
@johnugeorge
@gaocegege 

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/common/14)

### CODE REVIEW

**Comment by jian-he** [comment]
For the UT,  since it requires other code to be available to write a full UT,   can it be added later ?

**Comment by richardsliu** [comment]
Can we add some unit tests?

**Comment by jian-he** [comment]
sure, let me do it

File: job_controller/status.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	JobRunningReason = "JobRunning"
+	// JobFailedReason is added in a job when it is failed.
+	JobFailedReason = "JobFailed"
+	// JobRestarting is added in a job when it is restarting.
+	JobRestartingReason = "JobRestarting"
+
+	// labels for pods and servers.
+	ReplicaTypeLabel  = "replica-type"
+	ReplicaIndexLabel = "replica-index"
+	labelGroupName      = "group-name"
```
Comment (gaocegege) [inline job_controller/status.go]: We define the label as private but it is not used in the package, should we define it as a global var or just delete it?

File: util/k8sutil/k8sutil.go [hunk 2]
```diff
@@ -121,3 +122,19 @@ func FilterPodCount(pods []*v1.Pod, phase v1.PodPhase) int32 {
 	}
 	return result
 }
+
+func GetTotalReplicas(replicas map[commonv1.ReplicaType]*commonv1.ReplicaSpec) int32 {
+	tfjobReplicas := int32(0)
```
Comment (terrytangyuan) [inline util/k8sutil/k8sutil.go]: Remove mentions of "tfjob" in this function

File: job_controller/job_controller.go [hunk 3]
```diff
@@ -72,7 +79,7 @@ type ControllerInterface interface {
 	CreateService(job interface{}, service *v1.Service) error
 
 	// DeleteService deletes the service
-	DeleteService(job interface{}, service *v1.Service) error
+	DeleteService(job interface{}, name string, namespace string) error
```
Comment (johnugeorge) [inline job_controller/job_controller.go]: Can we make function signature consistent with CreateService?

Comment (gaocegege) [reply job_controller/job_controller.go]: Personally, I think we do not need to compose a service object to delete it.

Comment (johnugeorge) [reply job_controller/job_controller.go]: Is service argument meant to have the entire Service Object?  What is the `job` argument for?

File: util/k8sutil/k8sutil.go (hunk 2 above)
Comment (jian-he) [reply util/k8sutil/k8sutil.go]: done

File: job_controller/job_controller.go (hunk 3 above)
Comment (jian-he) [reply job_controller/job_controller.go]: the job argument is for the purpose of logging with job contexts such as its name
e.g  the r.Recorder.Eventf takes in the job as the argument

File: job_controller/status.go (hunk 1 above)
Comment (jian-he) [reply job_controller/status.go]: This is used in the newly added TestJobController

**Comment by jian-he** [comment]
Pushed a new change.
Added a TestJobController which implements the common interfaces and also the associated UTs around it. 
Fixed the other comments too

@gaocegege @richardsliu @terrytangyuan @johnugeorge

File: job_controller/status_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"testing"
+
+	"github.com/kubeflow/common/operator/v1"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestIsSucceeded(t *testing.T) {
```
Comment (richardsliu) [inline job_controller/status_test.go]: TestIsFailed?

File: job_controller/status_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		conditionInStatus := jobStatus.Conditions[0]
+		assert.Equal(t, conditionInStatus.Type, conditionType)
+		assert.Equal(t, conditionInStatus.Reason, reason)
+		assert.Equal(t, conditionInStatus.Message, message)
+	}
+}
+
+func TestUpdateJobReplicaStatuses(t *testing.T) {
+	jobStatus := v1.JobStatus{}
+	initializeReplicaStatuses(&jobStatus, "worker")
```
Comment (richardsliu) [inline job_controller/status_test.go]: Add asserts here for the initialized statuses

File: job_controller/status.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		condition.LastTransitionTime = currentCond.LastTransitionTime
+	}
+
+	// Append the updated condition to the conditions
+	newConditions := filterOutCondition(status.Conditions, condition.Type)
+	status.Conditions = append(newConditions, condition)
+}
+
+// filterOutCondition returns a new slice of job conditions without conditions with the provided type.
+func filterOutCondition(conditions []common.JobCondition, condType common.JobConditionType) []common.JobCondition {
```
Comment (richardsliu) [inline job_controller/status.go]: Is this tested?

File: job_controller/status.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+			return &condition
+		}
+	}
+	return nil
+}
+
+// setCondition updates the job to include the provided condition.
+// If the condition that we are about to add already exists
+// and has the same status and reason then we are not going to update.
+func setCondition(status *common.JobStatus, condition common.JobCondition) {
```
Comment (richardsliu) [inline job_controller/status.go]: Should have a unit test for this.

Comment (jian-he) [reply job_controller/status.go]: This is called by updateJobConditions and already covered by TestUpdateJobConditions

File: job_controller/status.go (hunk 6 above)
Comment (jian-he) [reply job_controller/status.go]: added

File: job_controller/status_test.go (hunk 4 above)
Comment (jian-he) [reply job_controller/status_test.go]: added

File: job_controller/status_test.go (hunk 5 above)
Comment (jian-he) [reply job_controller/status_test.go]: added

**Comment by jian-he** [comment]
@richardsliu check the new commit

**Pull request merged**